### PR TITLE
Add string xor modifier support

### LINKED ISSUE #51: Research the usefulness of accelerating xor modifier
Accelerate rules like:

```
rule XorExample1
{
    strings:
        $xor_string = "This program cannot" xor

    condition:
       $xor_string
}
```

My gut feeling is that this may result in way too many false positives, but I may be wrong.

We should probably transform strings like this to an OR of 256 xored strings in the backend. I don't feel like first-class support in ursadb is necessary (unless someone finds a way to optimise this that needs it).

### PR DESCRIPTION
**Pull request by nazywam**
**Your checklist for this pull request**
- [x] I've read the [contributing guideline](https://github.com/CERT-Polska/mquery/blob/master/CONTRIBUTING.md).
- [x] I've tested my changes by building and running mquery, and testing changed functionality (if applicable)
- [ ] I've added automated tests for my change (if applicable, optional)
- [ ] I've updated documentation to reflect my change (if applicable)

**What is the current behaviour?**

Yara strings with a `xor` modifiers are ignored

**What is the new behaviour?**

The yara->ursa parser replaces the xor-modified string with an `or` expression containing all possible cases

**Test plan**

**Closing issues**

fixes #51

### CODE REVIEW

**Comment by nazywam** [comment]
I've ran into some issues with the `xor` modifiers.
It turns out that yara allows a sub-modifiers for `xor` like this:

* `"hello world" xor(0x1f-0x20)`
* `"hello world" xor(34)`

They are parsed correctly by yaramod (version upgrade was needed) but as far as I see there is no way to read the using the python bindings: [code](https://github.com/avast/yaramod/blob/master/src/python/yaramod_python.cpp#L226)

One way of resolving this issue if parsing the `modifiers_text` ourselves but given the possibility of providing the xor range using both decimal and hex I'd really want to avoid this.

@msm-code can you weigh in on this?

**Comment by msm-code** [comment]
I think you're right that there is currently no way to get the modifiers out of the python bindings.

I think right now we can ignore the modifiers, and maybe we can submit a PR (or issue) upstream to fix this. After a few weeks it should land in pypi and we'll use this feature in mquery too.

Thoughts?

**Comment by nazywam** [comment]
>  we can submit a PR (or issue) upstream to fix

+1

Always wanted to learn how those binding thingies worked

File: src/lib/yaraparse.py [hunk 1]
```diff
@@ -148,11 +154,30 @@ def ursify_plain_string(string: PlainString) -> UrsaExpression:
...
+def ursify_xor_string(string: PlainString) -> UrsaExpression:
+    text_ascii = string.pure_text
+    xored_strings: List[UrsaExpression] = []
+
+    for xor_key in range(256):
+        xored_ascii = xor(text_ascii, bytes([xor_key]))
+        xored_wide = bytes(x ^ xor_key for y in text_ascii for x in [y, 0])
+
+        if string.is_ascii:
+            xored_strings.append(UrsaExpression(f"{{{xored_ascii.hex()}}}"))
```
Comment (msm-code) [inline src/lib/yaraparse.py:166]: created an issue to refactor it later #99 (also it's a good candidate for a first issue for someone)

File: src/lib/yaraparse.py [hunk 2]
```diff
@@ -148,11 +154,30 @@ def ursify_plain_string(string: PlainString) -> UrsaExpression:
         return ursa_ascii
 
 
+def ursify_xor_string(string: PlainString) -> UrsaExpression:
+    text_ascii = string.pure_text
+    xored_strings: List[UrsaExpression] = []
+
+    for xor_key in range(256):
```
Comment (msm-code) [inline src/lib/yaraparse.py:161]: Maybe add a comment like

```
# TODO(#100) implement modifier ranges

File: src/lib/yaraparse.py [hunk 3]
```diff
@@ -148,11 +154,30 @@ def ursify_plain_string(string: PlainString) -> UrsaExpression:
...
+        if string.is_wide:
+            xored_strings.append(UrsaExpression(f"{{{xored_wide.hex()}}}"))
+
+    return UrsaExpression.or_(*xored_strings)
+
+
 def ursify_string(string: String) -> Optional[UrsaExpression]:
-    if string.is_xor or string.is_nocase:
+    # nocase strings are not currently supported
+    if string.is_nocase:
```
Comment (msm-code) [inline src/lib/yaraparse.py:175]: at_least_not_yet

File: src/lib/yaraparse.py (hunk 2 above)
Comment (msm-code) [reply src/lib/yaraparse.py:161]: (#100 is the issue number. Not sure about the reference, we can also use a full link to be more explicit)

**Pull request merged**